relay-server/src/actors/project_cache.rs

```diff
@@ -40,6 +49,171 @@ pub enum ProjectError {
+            timestamp: UnixTimestamp::now().as_datetime(), // as good as any timestamp
+            scoping,
+            outcome,
+            event_id: None,
+            remote_addr: None,
+            category: DataCategory::TransactionProcessed,
+            quantity: transaction_count as u32,
+        });
+    }
+}
+
+/// Remove rate limited metrics buckets and track outcomes for them.
+///
+/// Returns any buckets that were *not* rate limited.
+async fn rate_limit_buckets(
+    buckets: Vec<Bucket>,
+    quotas: Vec<Quota>,
+    cached_rate_limits: RateLimits,
+    scoping: Scoping,
+    check_redis: bool,
+) -> Result<Vec<Bucket>, ()> {
+    // Collect bucket information (bucket, transaction_count)
+    let mut annotated_buckets: Vec<_> = buckets
+        .into_iter()
+        .map(|bucket| {
+            let mri = match MetricResourceIdentifier::parse(bucket.name.as_str()) {
+                Ok(mri) => mri,
+                Err(_) => {
+                    relay_log::error!("Invalid MRI: {}", bucket.name);
+                    return (bucket, None);
+                }
+            };
+
+            // Keep all metrics that are not transaction related:
+            if mri.namespace != MetricNamespace::Transactions {
+                // Not limiting sessions for now.
+                return (bucket, None);
+            }
+
+            if mri.name == "duration" {
+                // The "duration" metric is extracted exactly once for every processed transaction,
+                // so we can use it to count the number of transactions.
+                let count = bucket.value.len();
+                (bucket, Some(count))
+            } else {
+                // For any other metric in the transaction namespace, we check the limit with quantity=0
+                // so transactions are not double counted against the quota.
+                (bucket, Some(0))
+            }
+        })
+        .collect();
+
+    // Accumulate the total transaction count:
+    let total_transaction_count = annotated_buckets.iter().fold(
+        None,
+        |acc, (_, transaction_count)| match transaction_count {
+            Some(count) => Some(acc.unwrap_or(0) + count),
+            None => acc,
+        },
+    );
+
+    // Check rate limits if necessary:
+    if let Some(transaction_count) = total_transaction_count {
+        let item_scoping = ItemScoping {
+            category: DataCategory::TransactionProcessed,
+            scoping: &scoping,
+        };
+
+        let rate_limits = check_rate_limits(
+            transaction_count,
+            &cached_rate_limits,
+            quotas,
+            item_scoping,
+            check_redis,
+        )
+        .await;
+
+        match rate_limits {
+            Ok(rate_limits) => {
+                // If a rate limit is active, discard transaction buckets.
+                if let Some(limit) = rate_limits.limit_for(DataCategory::TransactionProcessed) {
```

Comment:
Is it even necessary to filter by data category here, given the `item_scoping`? Should we use `rate_limits.longest()` instead?